functionary/prompt.py

```diff
@@ -31,147 +101,191 @@ def from_message(cls, message: Dict) -> EndToken:
+            result["function_call"] = {
+                "name": func_name,
+                "arguments": arguments,
+            }  # FunctionCall(name=func_name, arguments=arguments)
+        if len(text_content) > 0:
+            result["content"] = text_content
+        return result
+
+
+class PromptTemplateV2(PromptTemplate):
+    from_token = "<|from|>"
+    recipient_token = "<|recipient|>"
+    content_token = "<|content|>"
+    stop_token = "<|stop|>"
+
+    def get_additional_tokens(self) -> List[str]:
+        return [
+            self.from_token,
+            self.recipient_token,
+            self.content_token,
+            self.stop_token,
+        ]
+
+    def get_text_from_message(self, message: Dict) -> str:
+        role = message["role"]
+        content = message.get("content", None)
 
-    elif message["role"] == "user" and content is None:
-        text = "user:\n"
+        if role in ["system", "user"]:  # <from>system\n<recipient>all\n<content>xxx
+            return f"{self.from_token}{role}\n{self.recipient_token}all\n{self.content_token}{content}\n"
 
-    elif message["role"] == "user":
-        text = f"user:\n{content}{end_token}\n"
+        if role == "tool":  # <from>tool_name\n<recipient>all\n<content>xxx
+            tool_name = message["name"]
+            return f"{self.from_token}{tool_name}\n{self.recipient_token}all\n{self.content_token}{content}\n"
 
-    elif message["role"] == "assistant":
+        assert role == "assistant"
+        tool_calls = message.get("tool_calls", [])
         if (
-            message.get("function_call", None) is not None
-        ):  # format of openai: {"role": assistant, "function_call": {"name": xxx, "arguments": xxx}}
-            function = message["function_call"]["name"]
-            arguments = message["function_call"]["arguments"] + end_token
-            if content is not None:
-                text = f"assistant:\n{content}\n{StartToken.function.value}{function}:\n{arguments}\n"
-            else:
-                text = f"assistant:\n{StartToken.function.value}{function}:\n{arguments}\n"
-        elif content is not None:  # this is text content
-            text = f"assistant:\n{content}{end_token}\n"
-        else:  # if no function call and content is None --> this is used at inference
-            text = "assistant:"
+            len(tool_calls) == 0 and content is None
+        ):  # for inference: <from> assistant\n<recipient>
+            return f"{self.from_token}{role}\n{self.recipient_token}"
 
-    return text
+        if len(tool_calls) == 0:  # <from>assistant\n<recipient>all\n<content>xxx
+            return f"{self.from_token}{role}\n{self.recipient_token}all\n{self.content_token}{content}{self.stop_token}\n"
 
+        result = ""
+        if content is not None:  # both text-response and function_call
+            result += f"{self.from_token}assistant\n{self.recipient_token}all\n{self.content_token}{content}\n"
```

Comment:
instead of hard coding assistant, we can use {role} as others

Reply:
Fixed